File: Windows/LineScope.Core/ImageOperation.cs
namespace LineScope.Core;

/// <summary>One step that derives a new image from an existing one.</summary>
public abstract record ImageOperation
{
    public sealed record Filter(FilterKind Kind) : ImageOperation;
    public sealed record Resample(ResampleMethod Method, double Scale) : ImageOperation;

    public string DisplayName => this switch
    {
        Filter f => f.Kind.DisplayName,
        Resample r => $"{r.Method.ShortName()} ×{Fmt.Number(r.Scale)}",
        _ => "",
    };

    public PixelBuffer Apply(PixelBuffer buffer) => this switch
    {
        Filter f => Filters.Apply(f.Kind, buffer),
        Resample r => Resampler.Resample(buffer, r.Scale, r.Method),
        _ => buffer,
    };
}

File: Packages/LineScopeCore/Sources/LineScopeCore/ImageOperation.swift
import Foundation

/// One step that derives a new image from an existing one.
public enum ImageOperation: Sendable, Equatable {
    case filter(FilterKind)
    case resample(ResampleMethod, scale: Double)

    public var displayName: String {
        switch self {
        case .filter(let f): f.displayName
        case .resample(let m, let s): "\(m.shortName) ×\(fmt(s))"
        }
    }

    public func apply(to buffer: PixelBuffer) -> PixelBuffer {
        switch self {
        case .filter(let f): Filters.apply(f, to: buffer)
        case .resample(let m, let s): Resampler.resample(buffer, scale: s, method: m)
        }
    }
}
